ProcessOptimizer/space/space.py

```diff
@@ -1,917 +1,896 @@
+        if any([isinstance(d, (str, bool)) or isinstance(d, np.bool_)
+                for d in dimension]):
+            return Categorical(dimension, transform=transform)
+        elif all([isinstance(dim, numbers.Integral) for dim in dimension]):
+            return Integer(*dimension, transform=transform)
+        elif any([isinstance(dim, numbers.Real) for dim in dimension]):
+            return Real(*dimension, transform=transform)
+        else:
+            raise ValueError("Invalid dimension {}. Read the documentation for"
+                             " supported types.".format(dimension))
+
+    if len(dimension) == 3:
+        if (any([isinstance(dim, (float, int)) for dim in dimension[:2]]) and
+                dimension[2] in ["uniform", "log-uniform"]):
+            return Real(*dimension, transform=transform)
+        else:
+            return Categorical(dimension, transform=transform)
+
+    if len(dimension) > 3:
+        return Categorical(dimension, transform=transform)
+
+    raise ValueError("Invalid dimension {}. Read the documentation for "
+                     "supported types.".format(dimension))
+
+
+class Dimension(ABC):
+    """Base class for search space dimensions."""
+
+    prior = None
+
+    def rvs(self, n_samples=1, random_state=None):
+        """Draw random samples.
+
+        Parameters
+        ----------
+        * `n_samples` [int or None]:
+            The number of samples to be drawn.
+
+        * `random_state` [int, RandomState instance, or None (default)]:
+            Set random state to something other than None for reproducible
+            results.
+        """
+        random_state = check_random_state(random_state)
+        index_array = random_state.uniform(size=n_samples)
+        return self.sample(index_array)
+
+    def transform(self, X):
+        """Transform samples form the original space to a warped space."""
+        return self.transformer.transform(X)
+
+    def inverse_transform(self, Xt):
+        """Inverse transform samples from the warped space back into the
+           original space.
+        """
+        return self.transformer.inverse_transform(Xt)
+
+    @property
+    def size(self):
+        return 1
+
+    @property
+    def transformed_size(self):
+        return 1
+
+    @property
+    def bounds(self):
+        raise NotImplementedError
+
+    @property
+    def transformed_bounds(self):
+        raise NotImplementedError
+
+    @property
+    def name(self):
+        return self._name
+
+    @name.setter
+    def name(self, value):
+        if isinstance(value, str) or value is None:
+            self._name = value
+        else:
+            raise ValueError("Dimension's name must be either string or None.")
+
+    def sample(
+        self, points: Union[int, float, Iterable[float]], deduplicate: bool = False
+    ) -> np.ndarray:
+        """Draw points from the dimension.
+
+        The sampling should be a reasonable mapping from the intercal [0, 1] to the
```

Comment:
Typo in "interval", and note that the implementation is on the interval [0, 1), that is 1 is not included. This was explicitly handled in the _uniform_inclusive function before (line 186), but I'm not really sure it matters with the rounding that will eventually happen when translating into specific factor settings.

Reply:
I have moved this description to the abstract method `_sample`, since it is an instruction for when you want to implement the abstract method.

It actually is [0,1] here.

rvs() only operates on [0,1), so it can now never give the upper bound of a real dimension. But the formal probability of that happening is 0 anyway, so I don't see that as a big issue. Do correct me if there is something I am missing.